File: best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if len(prices) == 0 or len(prices) == 1:
            return 0
        dp = [0, max(prices[1]-prices[0], 0)]
        minPrice = min(prices[0],prices[1])
        for i in range(2, len(prices)):
            dp += [ max(dp[i-1], prices[i] - minPrice) ]
            minPrice = min(minPrice, prices[i])
        return dp[-1]

File: best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.ts
function maxProfit(prices: number[]): number {
  let minPrice = Infinity;
  let maxProfit = 0;
  for (let i = 0; i < prices.length; i++) {
    let profit = prices[i] - minPrice;
    if (prices[i] < minPrice) minPrice = prices[i];
    else if (profit > maxProfit) maxProfit = profit;
  }
  return maxProfit;
}
